main.go

```diff
@@ -0,0 +1,169 @@
+	"os/signal"
+	"regexp"
+	"time"
+
+	"github.com/mattermost/mattermost-server/v6/model"
+	"github.com/rs/zerolog"
+)
+
+func main() {
+
+	app := &application{
+		logger: zerolog.New(
+			zerolog.ConsoleWriter{
+				Out:        os.Stdout,
+				TimeFormat: time.RFC822,
+			},
+		).With().Timestamp().Logger(),
+	}
+
+	app.config = loadConfig()
+	app.logger.Info().Str("config", fmt.Sprint(app.config)).Msg("")
+
+	setupGracefulShutdown(app)
+
+	// Create a new mattermost client.
+	app.mattermostClient = model.NewAPIv4Client(app.config.mattermostServer.String())
+
+	// Login.
+	app.mattermostClient.SetToken(app.config.mattermostToken)
+
+	if user, resp, err := app.mattermostClient.GetUser("me", ""); err != nil {
+		app.logger.Fatal().Err(err).Msg("Could not log in")
+	} else {
+		app.logger.Debug().Interface("user", user).Interface("resp", resp).Msg("")
+		app.logger.Info().Msg("Logged in to mattermost")
+		app.mattermostUser = user
+	}
+
+	// Find and save the bot's team to app struct.
+	if team, resp, err := app.mattermostClient.GetTeamByName(app.config.mattermostTeamName, ""); err != nil {
+		app.logger.Fatal().Err(err).Msg("Could not find team. Is this bot a member ?")
+	} else {
+		app.logger.Debug().Interface("team", team).Interface("resp", resp).Msg("")
+		app.mattermostTeam = team
+	}
+
+	// Find and save the talking channel to app struct.
+	if channel, resp, err := app.mattermostClient.GetChannelByName(
+		app.config.mattermostChannel, app.mattermostTeam.Id, "",
+	); err != nil {
+		app.logger.Fatal().Err(err).Msg("Could not find channel. Is this bot added to that channel ?")
+	} else {
+		app.logger.Debug().Interface("channel", channel).Interface("resp", resp).Msg("")
+		app.mattermostChannel = channel
+	}
+
+	// Send a message (new post).
+	sendMsgToTalkingChannel(app, "Hi! I am a bot.", "")
+
+	// Listen to live events coming in via websocket.
+	listenToEvents(app)
+}
+
+func setupGracefulShutdown(app *application) {
+	c := make(chan os.Signal, 1)
+	signal.Notify(c, os.Interrupt)
+	go func() {
+		for range c {
+			if app.mattermostWebSocketClient != nil {
+				app.logger.Info().Msg("Closing websocket connection")
+				app.mattermostWebSocketClient.Close()
+			}
+			app.logger.Info().Msg("Shutting down")
+			os.Exit(0)
+		}
+	}()
+}
+
+func sendMsgToTalkingChannel(app *application, msg string, replyToId string) {
+	// Note that replyToId should be empty for a new post.
+	// All replies in a thread should reply to root.
+
+	post := &model.Post{}
+	post.ChannelId = app.mattermostChannel.Id
+	post.Message = msg
+
+	post.RootId = replyToId
+
+	if _, _, err := app.mattermostClient.CreatePost(post); err != nil {
+		app.logger.Error().Err(err).Str("RootID", replyToId).Msg("Failed to create post")
+	}
+}
+
+func listenToEvents(app *application) {
+	var err error
+	failCount := 0
+	for {
+		app.mattermostWebSocketClient, err = model.NewWebSocketClient4(
+			fmt.Sprintf("ws://%s", app.config.mattermostServer.Host+app.config.mattermostServer.Path),
```

Comment:
Maybe it is a better best practice to use wss (Web Socket Secure) by default? https://www.rfc-editor.org/rfc/rfc6455#page-55

Reply:
Someone doing local testing with the docker-compose in this repo won't have SSL on their local machine.

Reply:
We could parse the Mattermost server URL and determine whether we are using `http` or `https`, and adjust the websocket protocol accordingly